cbmc_viewer/clauset.py

```diff
@@ -0,0 +1,263 @@
+            self.__dict__, VALID_CLAUSE_SUMMARY
+        )
+
+    def dump(self, filename=None, outdir=None):
+        """Write byteop metrics to a file or stdout."""
+
+        util.dump(self, filename, outdir)
+
+    def render_report(self, filename=None, outdir=None):
+        """Write the clause summary to a file rendered as html."""
+
+        clause_html = templates.render_clause_summary(self.summary,
+                                                      self.core,
+                                                      self.instrNotInCore)
+        util.dump(clause_html, filename or "clause.html", outdir)
+
+################################################################
+# Json key tags to access elements in cbmc json output
+
+JSON_GOTO_ID_KEY = 'GOTO_ID'
+JSON_SOURCE_LOC_KEY = 'Source'
+JSON_SAT_HARDNESS_KEY = 'SAT_hardness'
+JSON_CLAUSES_KEY = 'Clauses'
+JSON_CLAUSE_SET_KEY = 'ClauseSet'
+
+# Example cbmc json output
+# [
+#   {
+#     "GOTO": "// Labels: __CPROVER_HIDE__CPROVER_initialize();",
+#     "GOTO_ID": 0,
+#     "SAT_hardness": {
+#       "ClauseSet": [ ],
+#       "Clauses": 0,
+#       "Literals": 0,
+#       "Variables": 0
+#     },
+#     "SSA_expr": "true",
+#     "Source": {}
+#   },
+#   ...
+#   {
+#     "GOTO": "ASSERT main::1::a[cast(1, signedbv[64])] < 10",
+#     "GOTO_ID": 95,
+#     "SAT_hardness": {
+#       "ClauseSet": [ 6, 7, 8 ... ],
+#       "Clauses": 134,
+#       "Literals": 459,
+#       "Variables": 67
+#     },
+#     "SSA_expr": "¬(main::1::a!0@1#3[1] ≥ 10)",
+#     "Source": {
+#       "file": "simple_main.c",
+#       "function": "main",
+#       "line": "21",
+#       "workingDirectory": "..."
+#     }
+#   }
+# ]
+
+################################################################
+# Utility functions to generate clause summary
+
+def hash(clause):
+    # returns string format of clause to be used as hash key
+    return ' '.join([str(lit) for lit in clause])
+
+def get_loc_not_in_core(summary, core_instr):
+    instrNotInCore = []
+    coreSourceList = []
+    for instr in summary:
+        if instr[JSON_GOTO_ID_KEY] in core_instr:
+            coreSourceList.append(instr[JSON_SOURCE_LOC_KEY])
+
+    for instr in summary:
+        if instr[JSON_SAT_HARDNESS_KEY][JSON_CLAUSES_KEY] > 0:
+            if instr[JSON_SOURCE_LOC_KEY] not in coreSourceList:
+                if instr[JSON_SOURCE_LOC_KEY] not in instrNotInCore:
+                    instrNotInCore.append(instr[JSON_SOURCE_LOC_KEY])
+
+    return instrNotInCore
+
+# core_file: file storing the UNSAT core
+# clause_hash: mapping between clauses and goto intruction ID
+# instr_length: total number of instructions in the CBMC json output
+def get_instr_in_core(core_file, clause_hash, instr_length):
+    core_clauses = [] # clauses in core; clause is a list of literals
+    instr_list = [] # list of goto ID of instructions in core
+    keys = clause_hash.keys()
+
+    try:
+        with open(core_file) as f:
```

Comment:
There are two sources of clauses.  One is the dimacs output that we have already parsed and used to generate the clause_hash mapping string-clause representations of clauses to instruction ids.  One is the kissat output of the unsatisfiable core.  Here was are scanning the unsatisfiable core to generate the list of instruction ids contributing to the core.